plugin/mlsdbonlinelocator.cpp

```diff
@@ -0,0 +1,268 @@
+        return false;
+    }
+    qCDebug(lcGeoclueMlsdbOnline) << "Sent request with data:" << json;
+    return true;
+}
+
+void MlsdbOnlineLocator::requestOnlineLocationFinished(QNetworkReply *reply)
+{
+    QString errorString;
+    if (reply->error() == QNetworkReply::NoError) {
+        QByteArray data = reply->readAll();
+        qCDebug(lcGeoclueMlsdbOnline) << "MLS response:" << data;
+        if (!readServerResponseData(data, &errorString)) {
+            emit error(errorString);
+        }
+    } else {
+        emit error(reply->errorString());
+    }
+    reply->deleteLater();
+}
+
+bool MlsdbOnlineLocator::readServerResponseData(const QByteArray &data, QString *errorString)
+{
+    QJsonParseError parseError;
+    QJsonDocument json = QJsonDocument::fromJson(data, &parseError);
+    if (parseError.error != QJsonParseError::NoError) {
+        *errorString = parseError.errorString();
+        return false;
+    }
+
+    if (!json.isObject()) {
+        *errorString = "JSON parse error: expected object at root level, not found in " + data;
+        return false;
+    }
+
+    QJsonObject obj = json.object();
+    QVariantMap location = obj.value("location").toVariant().toMap();
+    if (location.isEmpty()) {
+        *errorString = "JSON parse error: no location data found in " + data;
+        return false;
+    }
+
+    bool latitudeOk = false;
+    bool longitudeOk = false;
+    double latitude = location["lat"].toDouble(&latitudeOk);
+    double longitude = location["lng"].toDouble(&longitudeOk);
+    if (!latitudeOk || !longitudeOk) {
+        *errorString = "JSON parse error: latitude or longitude not readable in " + data;
+        return false;
+    }
+
+    bool accuracyOk = false;
+    double accuracy = obj.value("accuracy").toVariant().toDouble(&accuracyOk);
+    if (!accuracyOk) {
+        accuracy = -1;
+    }
+    emit locationFound(latitude, longitude, accuracy);
+    return true;
+}
+
+QVariantMap MlsdbOnlineLocator::globalFields()
+{
+    QVariantMap map;
+    if (!m_simManager || !m_simManager->isValid()) {
+        return map;
+    }
+    map["carrier"] = m_simManager->serviceProviderName();
+    map["considerIp"] = true;
+    map["homeMobileCountryCode"] = m_simManager->mobileCountryCode();
+    map["homeMobileNetworkCode"] = m_simManager->mobileNetworkCode();
+    return map;
+}
+
+QVariantMap MlsdbOnlineLocator::cellTowerFields(const QList<MlsdbProvider::CellPositioningData> &cells)
+{
+    QVariantMap map;
+    QVariantList cellTowers;
+    Q_FOREACH (const MlsdbProvider::CellPositioningData &cell, cells) {
+        QVariantMap cellTowerMap;
+        // supported radio types: gsm, wcdma or lte
+        switch (cell.uniqueCellId.cellType()) {
+        case MLSDB_CELL_TYPE_LTE:
+            cellTowerMap["radioType"] = "lte";
+            break;
+        case MLSDB_CELL_TYPE_GSM:
+            cellTowerMap["radioType"] = "gsm";
+            break;
+        default:
```

Comment:
In the MLS database, WCDMA are reported as UMTS I believe.  So you can do case MLSDB_CELL_TYPE_UMTS: cellTowerMap["radioType"] = "wcdma"; break; I believe.